qutip/core/data/matmul.pyx

```diff
@@ -289,6 +290,61 @@ cpdef Dense matmul_dense(Dense left, Dense right, double complex scale=1, Dense
     return out
 
 
+cpdef CSR element_wise_multiply_csr(CSR left, CSR right):
+    """Element-wise multiplication of CSR matrices."""
+    if left.shape[0] != right.shape[0] or left.shape[1] != right.shape[1]:
+        raise ValueError(
+            "incompatible matrix shapes "
+            + str(left.shape)
+            + " and "
+            + str(right.shape)
+        )
+    cdef idxint col_left, left_nnz = csr.nnz(left)
+    cdef idxint col_right, right_nnz = csr.nnz(right)
+    cdef idxint ptr_left, ptr_right, ptr_left_max, ptr_right_max
+    cdef idxint row, nnz=0, ncols=left.shape[1]
+    cdef CSR out
+    # Fast paths for zero matrices.
+    if right_nnz == 0 or left_nnz == 0:
+        return csr.zeros(left.shape[0], left.shape[1])
+    # Main path.
+    out = csr.empty(left.shape[0], left.shape[1], max(left_nnz, right_nnz))
```

Comment:
Assuming we have to support nans, some of this logic will fail. E.g. `[nan, nan] * [0, 0] -> [nan, nan]` even though `right_nnz == 0` and similarly `[nan, 0] * [0, nan] -> [nan, nan]` which has two non-zero elements even though `max(left_nnz, right_nnz) == 1`.